### DISCUSSION TITLE
CPU and GPU useage

### DISCUSSION THREAD

**Discussion by xland** [mikke89/RmlUi#331] [General]
I know RmlUi is a immediately gui framework,like Dear ImGui(is it true?).
But,RmlUi uses more resources than ImGui.
Will you make some improvements in this area in the future?
![image](https://user-images.githubusercontent.com/2980915/175855555-1a27c51f-2482-4092-bc17-c2f9ea47182a.png)

**Comment by mikke89** [comment]
RmlUi is a retained mode UI unlike ImGui which is immediate mode. Nevertheless, performance and resource usage is a much larger topic than the choice between these two modes alone.

My impression is that neither ImGui or RmlUi are optimized in terms of idle usage. Thus comparisons like this are very unreliable as they mainly depend on the backend used, your GPU driver, GPU settings, and whether vsync is enabled in particular. For example, on my end I get substantially better performance in RmlUi:

![image](https://user-images.githubusercontent.com/5490330/175906494-9d91b9d9-7214-46af-bb4a-56ebc7fae30b.png)

The GPU usage you show for ImGui seems like a result of the task manager just plainly not working for DirectX 12. Try e.g. DirectX 11 instead and you will see that it shows up there.

**Reply by xland** [reply]
Thanks for your reply.
Yes,This is not a good test.
I believe RmlUi will definitely get better.
I decided to ignore it and focus on my logic development work.

**Reply by mikke89** [reply]
What you are probably looking for is a power saving feature. I see there are a few open PRs for this for ImGui that we could borrow ideas from: https://github.com/ocornut/imgui/pull/2749 and https://github.com/ocornut/imgui/pull/5116. This is not a priority for me but maybe someone else wants to tackle something like this for RmlUi.

**Comment by xland** [comment]
Why doesn't RmlUi stay at 0% CPU useage on process idle?
Does OpenGL cause that useage?
Thanks~

**Reply by mikke89** [reply]
Currently the backends don't do anything special to limit CPU usage, they just loop naively. See my comment about power saving feature above.
